torch/multiprocessing/reductions.py

```diff
@@ -336,6 +343,48 @@ def reduce_tensor(tensor):
     return (rebuild_tensor, (type(tensor), storage, metadata))
 
 
+def rebuild_nested_tensor(
+    rebuild_buffer_func,
+    rebuild_buffer_args,
+    rebuild_sizes_func,
+    rebuild_sizes_args,
+    rebuild_strides_func,
+    rebuild_strides_args,
+    rebuild_offsets_func,
+    rebuild_offsets_args,
+):
+    buffer = rebuild_buffer_func(*rebuild_buffer_args)
+    sizes = rebuild_sizes_func(*rebuild_sizes_args)
+    strides = rebuild_strides_func(*rebuild_strides_args)
+    offsets = rebuild_offsets_func(*rebuild_offsets_args)
+    return torch._nested_view_from_buffer(buffer, sizes, strides, offsets)
+
+
+def reduce_nested_tensor(nt):
+    rebuild_buffer_func, rebuild_buffer_args = reduce_tensor(nt.values().clone())
```

Comment:
Why clone() ?

Reply:
Without the clone():
```
RuntimeError: Attempted to send CUDA tensor received from another process; this is not currently supported. Consider cloning before sending.
```

Reply:
When you do what?

Reply:
When serializing, `reduce_tensor()` is called on the NT buffer, and `_share_cuda()` for the storage breaks:
```
Traceback (most recent call last):
  File "/home/jbschlosser/anaconda3/lib/python3.9/multiprocessing/queues.py", line 245, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/home/jbschlosser/anaconda3/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "/home/jbschlosser/branches/nt_ops/torch/multiprocessing/reductions.py", line 298, in reduce_tensor
    return reduce_nested_tensor(tensor)
  File "/home/jbschlosser/branches/nt_ops/torch/multiprocessing/reductions.py", line 364, in reduce_nested_tensor
    rebuild_buffer_func, rebuild_buffer_args = reduce_tensor(nt.values())
  File "/home/jbschlosser/branches/nt_ops/torch/multiprocessing/reductions.py", line 310, in reduce_tensor
    ) = storage._share_cuda_()
  File "/home/jbschlosser/branches/nt_ops/torch/storage.py", line 1073, in _share_cuda_
    return self._untyped_storage._share_cuda_(*args, **kwargs)
RuntimeError: Attempted to send CUDA tensor received from another process; this is not currently supported. Consider cloning before sending.
```

Reply:
But is that because you try to send stuff that you received from another process as the error suggests? :p

Reply:
Haha probably, but I'm unclear as to why that's happening. This is from the test case in this PR that operates with `batch_size=1` and `num_workers>0`. I'd expect each NT in the dataset to be sent only once to one of the workers, but something more complicated than this is happening.

Reply:
Okay I figured it out: the worker threads are sending stuff they received from the parent thread due to the use of `_identity()`. I'm going to remove the `clone()` and fix the test case.